RecoHGCal/TICL/plugins/TrackstersFromCaloParticlesProducer.cc

```diff
@@ -0,0 +1,130 @@
+// Author: Leonardo Cristella - leonardo.cristella@cern.ch
+// Date: 08/2021
+
+// user include files
+
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "DataFormats/CaloRecHit/interface/CaloCluster.h"
+#include "DataFormats/ParticleFlowReco/interface/PFCluster.h"
+
+#include "DataFormats/HGCalReco/interface/Trackster.h"
+
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "SimDataFormats/Associations/interface/LayerClusterToCaloParticleAssociator.h"
+
+#include "SimDataFormats/CaloAnalysis/interface/CaloParticle.h"
+#include "RecoLocalCalo/HGCalRecAlgos/interface/RecHitTools.h"
+
+#include "RecoHGCal/TICL/interface/commons.h"
+
+#include "TrackstersPCA.h"
+#include <vector>
+#include <iterator>
+#include <algorithm>
+
+using namespace ticl;
+
+class TrackstersFromCaloParticlesProducer : public edm::stream::EDProducer<> {
+public:
+  explicit TrackstersFromCaloParticlesProducer(const edm::ParameterSet&);
+  ~TrackstersFromCaloParticlesProducer() override {}
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+private:
+  std::string detector_;
+  const bool doNose_ = false;
+  const edm::EDGetTokenT<std::vector<reco::CaloCluster>> clusters_token_;
+  const edm::EDGetTokenT<edm::ValueMap<std::pair<float, float>>> clustersTime_token_;
+  const edm::EDGetTokenT<std::vector<float>> filtered_layerclusters_mask_token_;
+
+  edm::EDGetTokenT<std::vector<CaloParticle>> caloparticles_token_;
+
+  edm::InputTag associatorLayerClusterCaloParticle_;
+  edm::EDGetTokenT<hgcal::SimToRecoCollection> associatorMapCaloParticleToReco_token_;
+  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> geom_token_;
+  hgcal::RecHitTools rhtools_;
+  const double fractionCut_;
+};
+DEFINE_FWK_MODULE(TrackstersFromCaloParticlesProducer);
+
+TrackstersFromCaloParticlesProducer::TrackstersFromCaloParticlesProducer(const edm::ParameterSet& ps)
+    : detector_(ps.getParameter<std::string>("detector")),
+      doNose_(detector_ == "HFNose"),
+      clusters_token_(consumes<std::vector<reco::CaloCluster>>(ps.getParameter<edm::InputTag>("layer_clusters"))),
```

Comment:
```suggestion
      clusters_token_(consumes(ps.getParameter<edm::InputTag>("layer_clusters"))),
```
types are not required; please use the simpler interface (same for the rest)